### PR TITLE
remove expired rows from cache on read

### LINKED ISSUE #6033: Expired tombstones are kept in cache indefinitely
*Installation details*
HEAD: 0cb718276
Cluster size: 1

*Description*
Tombstones seems to be kept in the cache even after 'nodetool flush'.

How to reproduce:
1) Create a 1 node ccm cluster:
```
ccm create scylla-1 --scylla --vnodes -n 1 --install-dir=<scylla repo>
```
2) Start it:
```
ccm start --jvm_arg="--memory" --jvm_arg="1G"  --jvm_arg="--smp" --jvm_arg="1"
```
3) From cqlsh create the table with `gc_grace_seconds=0`:
```
CREATE KEYSPACE ks2 WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}  AND durable_writes = true;

CREATE TABLE ks2.test (
    a int PRIMARY KEY,
    b text,
    vladz1 int,
    vladz2 int
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 0
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

```
4) Run the test where we create and entry, delete it with the tombstone in the future, then call `flush` and `compact` and still unable to read the re-inserted entry until we restart scylla:
```
fred@cqlsh> INSERT INTO ks2.test (a, vladz1) VALUES (1, 3);                                                                                                                                                                                                                                                                   
fred@cqlsh> INSERT INTO ks2.test (a, vladz1) VALUES (2, 4);                                                                                                                                                                                                                                                                   
fred@cqlsh> SELECT * FROM ks2.test;

 a | b    | vladz1 | vladz2
---+------+--------+--------
 1 | null |      3 |   null
 2 | null |      4 |   null

(2 rows)
fred@cqlsh> DELETE FROM ks2.test USING timestamp 9984457162684266 where a=1;                                                                                                                                                                                                                                                  
fred@cqlsh> SELECT * FROM ks2.test;                                                                                                                                                                                                                                                                                           

 a | b    | vladz1 | vladz2
---+------+--------+--------
 2 | null |      4 |   null

(1 rows)
fred@cqlsh> INSERT INTO ks2.test (a, vladz1) VALUES (1, 3);
fred@cqlsh> SELECT * FROM ks2.test;

 a | b    | vladz1 | vladz2
---+------+--------+--------
 2 | null |      4 |   null

(1 rows)
fred@cqlsh> exit
$ ../scylla-tools-java/bin/nodetool -p 7100 flush
$ ../scylla-tools-java/bin/nodetool -p 7100 compact
$ ~/.ccm/scylla-1/node1/data/ks2/test-83c06e70686a11ea8c6f000000000000/
total 40K
drwxrwxr-x 5 vladz vladz  303 Mar 17 12:16 .
drwxrwxr-x 3 vladz vladz   51 Mar 17 12:15 ..
-rw-rw-r-- 1 vladz vladz   66 Mar 17 12:16 mc-2-big-CompressionInfo.db
-rw-rw-r-- 1 vladz vladz   35 Mar 17 12:16 mc-2-big-Data.db
-rw-rw-r-- 1 vladz vladz   10 Mar 17 12:16 mc-2-big-Digest.crc32
-rw-rw-r-- 1 vladz vladz  176 Mar 17 12:16 mc-2-big-Filter.db
-rw-rw-r-- 1 vladz vladz    8 Mar 17 12:16 mc-2-big-Index.db
-rw-rw-r-- 1 vladz vladz   78 Mar 17 12:16 mc-2-big-Scylla.db
-rw-rw-r-- 1 vladz vladz 4.6K Mar 17 12:16 mc-2-big-Statistics.db
-rw-rw-r-- 1 vladz vladz   56 Mar 17 12:16 mc-2-big-Summary.db
-rw-rw-r-- 1 vladz vladz  102 Mar 17 12:16 mc-2-big-TOC.txt
drwxrwxr-x 2 vladz vladz    6 Mar 17 12:16 pending_delete
drwxrwxr-x 2 vladz vladz    6 Mar 17 12:15 staging
drwxrwxr-x 2 vladz vladz    6 Mar 17 12:15 upload
$ ~/work/scylla-tools-java/bin/cqlsh
Connected to scylla-1 at 127.0.0.1:9042.
[cqlsh 5.0.1 | Cassandra 3.0.8 | CQL spec 3.3.1 | Native protocol v4]
Use HELP for help.
fred@cqlsh> INSERT INTO ks2.test (a, vladz1) VALUES (1, 3);
fred@cqlsh> SELECT * FROM ks2.test;

 a | b    | vladz1 | vladz2
---+------+--------+--------
 2 | null |      4 |   null

(1 rows)
fred@cqlsh> exit

<Here we restart scylla and log into cqlsh again>

$ ~/work/scylla-tools-java/bin/cqlsh 
Connected to scylla-1 at 127.0.0.1:9042.
[cqlsh 5.0.1 | Cassandra 3.0.8 | CQL spec 3.3.1 | Native protocol v4]
Use HELP for help.
fred@cqlsh> SELECT * FROM ks2.test;

 a | b    | vladz1 | vladz2
---+------+--------+--------
 1 | null |      3 |   null
 2 | null |      4 |   null

(2 rows)
```

## Current status
### Tombstones covered so far:

- [x] Range tombstones: https://github.com/scylladb/scylladb/commit/87fa7d0381943ef3404a1d630baf00803b2800f5
- [ ] Partition tombstones
- [x] Row tombstones: https://github.com/scylladb/scylladb/issues/6033#issuecomment-1703398546
- [ ] Cell tombstones

### PR DESCRIPTION
**Pull request by alezzqz**
- check and remove expired tombstones on read from row cache
- prevent populating cache with expired rows from sstables (compact on read) - maybe we should apply this in separate patch?

there are few questions in code comments

Refs #2252
Fixes #6033

### CODE REVIEW

File: cache_flat_mutation_reader.hh [hunk 1]
```diff
@@ -642,10 +657,38 @@ void cache_flat_mutation_reader::copy_from_cache_to_buffer() {
     }, _read_context.is_reversed()) == stop_iteration::no) {
         return;
     }
-    // We add the row to the buffer even when it's full.
-    // This simplifies the code. For more info see #3139.
+
     if (_next_row_in_range) {
-        add_to_buffer(_next_row);
```
Comment (tgrabiec) [inline cache_flat_mutation_reader.hh:648]: The patch shouldn't close https://github.com/scylladb/scylladb/issues/6033 because it doesn't deal with all tombstones. It doesn't expire range tombstones, for example.

I don't think we should try to handle range tombstones before https://github.com/scylladb/scylladb/pull/12048 is merged, which makes significant changes to the structure of cache with regards to range tombstones and MVCC.

File: cache_flat_mutation_reader.hh [hunk 2]
```diff
@@ -642,10 +657,38 @@ void cache_flat_mutation_reader::copy_from_cache_to_buffer() {
...
+        bool remove_row = false;
+
+        if (_state == state::reading_from_cache && !_next_row.dummy()) { // maybe we'd better use at_a_row() instead of dummy() here?
+            deletable_row& row = _next_row.latest_row();
```
Comment (tgrabiec) [inline cache_flat_mutation_reader.hh:665]: `_state == state::reading_from_cache` is guaranteed here.

File: cache_flat_mutation_reader.hh [hunk 3]
```diff
@@ -642,10 +657,38 @@ void cache_flat_mutation_reader::copy_from_cache_to_buffer() {
...
+            tombstone base_tombstone = _snp->version()->partition().range_tombstone_for_row(*_schema, _next_row.key());
```
Comment (tgrabiec) [inline cache_flat_mutation_reader.hh:666]: That will be very expensive, the price is too high to pay per-row.

We should either wait for https://github.com/scylladb/scylladb/pull/12048, or make this compaction weaker by not considering range tombstones. 

You could ignore base_tombstone in is_live() since we're not really interested

File: cache_flat_mutation_reader.hh [hunk 4]
```diff
@@ -642,10 +657,38 @@ void cache_flat_mutation_reader::copy_from_cache_to_buffer() {
...
+
+            if (!row.is_live(*_schema, column_kind::regular_column, base_tombstone, _read_time) && // should we really call is_live here?
+                row.deleted_at().max_deletion_time() < _gc_before) {
+                remove_row = true;
```
Comment (tgrabiec) [inline cache_flat_mutation_reader.hh:670]: !row.is_live() is not the correct check. Row can be not live but still have not expired tombstones inside cells, which we must not drop.

The proper way to do it is to compact it using row.compact_and_expire() like mutation_compactor does, and then drop the row only if it is empty().

File: cache_flat_mutation_reader.hh [hunk 5]
```diff
@@ -642,10 +657,38 @@ void cache_flat_mutation_reader::copy_from_cache_to_buffer() {
...
+
+        if (remove_row) {
+            with_allocator(_snp->region().allocator(), [&] {
+                partition_snapshot_row_weakref row_ref(_next_row);
+
+                cache_tracker& tracker = _read_context.cache()._tracker;
+                if (row_ref->is_linked()) {
+                    tracker.get_lru().remove(*row_ref);
+                }
+                row_ref->on_evicted(tracker);
```
Comment (tgrabiec) [inline cache_flat_mutation_reader.hh:682]: You cannot evict from latest version if there are older MVCC versions, see https://github.com/scylladb/scylladb/blob/master/docs/dev/row_cache.md:

```
Another rule is that row entries in older versions are evicted first, before any row in the newer version is evicted. This is needed so that we don't appear to loose writes in case we have the same row in more than one version, and the one from the newer version gets evicted first. To achieve this, we only move to the front of the LRU (marking as more recently used, evicted last) row entries which belong to the latest partition_version in a given partition_entry. This implies that detached snapshots never update the LRU.
```

There is already `cache_flat_mutation_reader::maybe_drop_last_entry()`, which checks for sufficient conditions for removal. Can we extend that?

File: replica/table.cc [hunk 6]
```diff
@@ -1470,7 +1470,7 @@ table::sstables_as_snapshot_source() {
             return make_compacting_reader(
                 std::move(reader),
                 gc_clock::now(),
-                [](const dht::decorated_key&) { return api::min_timestamp; },
+                [](const dht::decorated_key&) { return api::max_timestamp; },
```
Comment (tgrabiec) [inline replica/table.cc:1473]: Should be a separate patch with its own description.

File: row_cache.hh [hunk 7]
```diff
@@ -362,15 +362,29 @@ public:
...
+                                     streamed_mutation::forwarding fwd = streamed_mutation::forwarding::no,
+                                     mutation_reader::forwarding fwd_mr = mutation_reader::forwarding::no) {
+        return make_reader(s, permit, range, slice, nullptr, pc, trace_state, fwd, fwd_mr);
+    }
+
     flat_mutation_reader_v2 make_reader(schema_ptr s,
                                      reader_permit permit,
                                      const dht::partition_range& range,
                                      const query::partition_slice& slice,
+                                     const tombstone_gc_state* gc_state,
```
Comment (tgrabiec) [inline row_cache.hh:382]: You could add this as a last parameter and default to nullptr. This way you don't need to add an overload.

File: row_cache.hh [hunk 8]
```diff
@@ -381,14 +395,18 @@ public:
...
                                      mutation_reader::forwarding fwd_mr = mutation_reader::forwarding::no);
 
-    flat_mutation_reader_v2 make_reader(schema_ptr s, reader_permit permit, const dht::partition_range& range = query::full_partition_range) {
+    flat_mutation_reader_v2 make_reader(schema_ptr s,
+                                    reader_permit permit,
+                                    const dht::partition_range& range = query::full_partition_range,
+                                    const tombstone_gc_state* gc_state = nullptr) {
         auto& full_slice = s->full_slice();
-        return make_reader(std::move(s), std::move(permit), range, full_slice);
+        return make_reader(std::move(s), std::move(permit), range, full_slice, gc_state);
```
Comment (tgrabiec) [inline row_cache.hh:409]: Unit test which demonstrates this feature in row_cache_test.cc is missing.

File: cache_flat_mutation_reader.hh (hunk 3 above)
Comment (alezzqz) [reply cache_flat_mutation_reader.hh:666]: I agree, that really could be expensive but don't know better solution in current implementation.
Do you think we can just ignore range tombstones?

File: replica/table.cc (hunk 6 above)
Comment (alezzqz) [reply replica/table.cc:1473]: Okay, will do it as separate patch. And this change leads `incremental_compaction_data_resurrection_test` to failure. We have to fix the test as well.

File: row_cache.hh (hunk 7 above)
Comment (alezzqz) [reply row_cache.hh:382]: It was the first thing I did but it doesn't help unfortunately: some tests pass own values and don't use default ones so either we make changes in many tests or overload the method. I chose the simplest way )

File: row_cache.hh (hunk 8 above)
Comment (alezzqz) [reply row_cache.hh:409]: Thanks, will do

File: cache_flat_mutation_reader.hh (hunk 3 above)
Comment (avikivity) [reply cache_flat_mutation_reader.hh:666]: I recommend to rebase your patches on top of #12048, it will allow everything to be considered.

File: cache_flat_mutation_reader.hh (hunk 1 above)
Comment (alezzqz) [reply cache_flat_mutation_reader.hh:648]: I see. Do you think #12048 is expected to be merged soon?

File: cache_flat_mutation_reader.hh (hunk 5 above)
Comment (alezzqz) [reply cache_flat_mutation_reader.hh:682]: Thanks, will check the subject

File: cache_flat_mutation_reader.hh (hunk 1 above)
Comment (tgrabiec) [reply cache_flat_mutation_reader.hh:648]: Yes. You can rebase on top of it as Avi suggested.

File: cache_flat_mutation_reader.hh (hunk 4 above)
Comment (alezzqz) [reply cache_flat_mutation_reader.hh:670]: you mean something like following?
```
bool expired(tombstone t) {
    return (t && t < _gc_before);
}

if (expired(range_tombstone) || expired(row_tombstone)) {
    if (!row.compact_and_expire(...)) {
        remove_row = true;
    }
}
```
anyway in some cases compact_and_expire (looks like expensive operation a little) will be called twice for each row: in `cache_flat_mutation_reader` and `mutation_compactor` and it makes me feel embarrassed

File: cache_flat_mutation_reader.hh (hunk 3 above)
Comment (alezzqz) [reply cache_flat_mutation_reader.hh:666]: Thanks, I did. However I could apply my changes (based on #12048) only when it will be finished

File: cache_flat_mutation_reader.hh (hunk 4 above)
Comment (tgrabiec) [reply cache_flat_mutation_reader.hh:670]: Something like that, except that remove_row should not depend on the result of compact_and_expire(), but row.empty(). compact_and_expire() returns "is live". !is_live doesn't mean that we can remove the row, because it may still contain non-expired tombstones. Only when it contains no information after compaction, so empty(), can we remove it.

Comment (tgrabiec) [reply cache_flat_mutation_reader.hh:670]:
```
    if (!row.compact_and_expire(...)) {
        remove_row = row.empty();
    }
```
